src/apps/external_executor/external_executor.cpp

```diff
@@ -466,6 +513,75 @@ namespace externalexecutor
         executor_only)
         .set_forwarding_required(ccf::endpoints::ForwardingRequired::Never)
         .install();
+
+      auto stream = [this](
+                      ccf::endpoints::EndpointContext& ctx,
+                      google::protobuf::Empty&& payload) {
+        // Dummy streaming endpoint
+
+        // TODO:
+        // 1. Create gRPC server stream wrapper that sets stream as non-unary
+        // 2. [DONE] Add ability to close stream from rpc_ctx->stream(data,
+        // close=true) or rpc_ctx->stream_close();
+        // 3. Create stream object from rpc_ctx, and then
+        // rpc_ctx->create_stream() (figure out ownership and lifetime)
+
+        LOG_FAIL_FMT("Endpoint synchronous execution");
+
+        auto stream =
+          ccf::grpc::make_stream<externalexecutor::protobuf::KVKeyValue>(
+            ctx.rpc_ctx, responder_lookup);
+        async_send_stream_data(stream);
+
+        // TODO: Fix return value here, which should return either another
+        // object or nothing
+        return ccf::grpc::make_success(
+          ccf::grpc::Stream<externalexecutor::protobuf::KVKeyValue>{});
+      };
+
+      make_endpoint(
+        "/externalexecutor.protobuf.KV/Stream",
+        HTTP_POST,
+        ccf::grpc_adapter<
+          google::protobuf::Empty,
+          ccf::grpc::Stream<externalexecutor::protobuf::KVKeyValue>>(stream),
+        {ccf::no_auth_required})
+        .install();
+
+      auto stream_inside = [this](
+                             ccf::endpoints::EndpointContext& ctx,
+                             google::protobuf::Empty&& payload) {
+        // TODO:
+        // 1. Figure out stream ownership and lifetime
+        // 2. Create detachable stream
+
+        auto stream =
+          ccf::grpc::make_stream<externalexecutor::protobuf::KVKeyValue>(
+            ctx.rpc_ctx, responder_lookup);
+
+        for (int i = 0; i < 5; i++)
+        {
+          bool close_stream = i == 4;
+          externalexecutor::protobuf::KVKeyValue kv = {};
+          kv.set_key("my_key");
+          kv.set_value(fmt::format("my_value: {}", close_stream));
+          stream->stream_msg(kv, close_stream);
+        }
+
+        // TODO: Fix return value here, which should return success but nothing
+        return ccf::grpc::make_success(
+          ccf::grpc::Stream<externalexecutor::protobuf::KVKeyValue>{});
```

Comment:
I think something like `return ccf::grpc::pending();` would actually be clearer than `ccf::grpc::make_success();`. But it may also be better to avoid these return-based wrappers for streaming endpoints? So `grpc_adapter` gives us 2 features; type-safety, and a transformation to stack-constrained API (here's your params, return your result). We want to keep the former, but the latter can be dropped for these if it makes sense.

Reply:
Done. The adapter is now much cleaner, and creates the `Stream` for the handler.